target-chains/near/receiver/src/governance.rs

```diff
@@ -34,162 +33,218 @@ use {
+    SetFee { base: u64, expo: u64 },
+    SetValidPeriod { valid_seconds: u64 },
+    RequestGovernanceDataSourceTransfer { governance_data_source_index: u32 },
 }
 
-impl GovernanceAction {
-    pub fn id(&self) -> ActionId {
-        match self {
-            GovernanceAction::ContractUpgrade(_) => ActionId::ContractUpgrade,
-            GovernanceAction::SetDataSources(_) => ActionId::SetDataSources,
-            GovernanceAction::SetGovernanceSource(_) => ActionId::SetGovernanceSource,
-            GovernanceAction::SetStalePriceThreshold(_) => ActionId::SetStalePriceThreshold,
-            GovernanceAction::SetUpdateFee(_) => ActionId::SetUpdateFee,
-        }
-    }
+#[derive(BorshDeserialize, BorshSerialize, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct GovernanceInstruction {
+    pub module: GovernanceModule,
+    pub action: GovernanceAction,
+    pub target: Chain,
+}
 
-    pub fn deserialize(data: &[u8]) -> Result<Self, Error> {
-        let mut cursor = std::io::Cursor::new(data);
-        let magic = cursor.read_u32::<BigEndian>()?;
-        let module = cursor.read_u8()?;
-        let action = cursor.read_u8()?.try_into()?;
-        let target = cursor.read_u16::<BigEndian>()?;
-
-        assert!(module == GOVERNANCE_MODULE);
-        assert!(target == 0 || target == u16::from(WormholeChain::Near));
-        assert!(magic == u32::from_le_bytes(GOVERNANCE_MAGIC));
-
-        Ok(match action {
-            ActionId::ContractUpgrade => {
-                let mut hash = [0u8; 32];
-                cursor.read_exact(&mut hash)?;
-                Self::ContractUpgrade(hash)
-            }
+impl GovernanceInstruction {
+    /// Implements a `deserialize` method for the `GovernanceAction` enum using `nom` to
+    /// deserialize the payload. The use of `nom` gives us parser safety, error handling, full
+    /// buffer consumption, and a more readable implementation while staying efficient.
+    pub fn deserialize(input: impl AsRef<[u8]>) -> Result<Self, Error> {
+        use nom::{
+            bytes::complete::take,
+            combinator::all_consuming,
+            multi::length_count,
+            number::complete::{
+                be_u16,
+                be_u32,
+                be_u64,
+                be_u8,
+            },
+        };
+
+        let input = input.as_ref();
+
+        // Verify Governance header is as expected so we can bail to avoid more parsing.
+        let (input, magic) = take(4usize)(input)?;
+        let (input, module) = be_u8(input)?;
+        let (input, action) = be_u8(input)?;
+        let (input, chain) = be_u16(input)?;
+        let module = GovernanceModule::from_u8(module).ok_or(InvalidGovernanceModule)?;
+        let chain = Chain::from(WormholeChain::from(chain));
+
+        // Safely parse the action ID. [ref:action_discriminants]
+        let action = GovernanceActionId::from_u8(action).ok_or(InvalidGovernanceAction)?;
```

Comment:
We should also check that the chain id NEAR or "all chains" (0).

Reply:
can you also add some tests to ensure it gets rejected in that case.